sale_margin_delivered/models/stock_picking.py

```diff
@@ -0,0 +1,20 @@
+# Copyright 2018 Tecnativa - Sergio Teruel
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import api, models
+
+
+class StockPicking(models.Model):
+    _inherit = "stock.picking"
+
+    @api.multi
+    def action_done(self):
+        res = super().action_done()
+        # Update purchase_price in so_lines
+        pickings = self.filtered(
+            lambda x: x.picking_type_code == 'outgoing')
+        so_lines = pickings.mapped(
+            'move_lines.sale_line_id')
+        for line in so_lines:
+            line.purchase_price = line.product_id.standard_price
```

Comment:
Why this code, @sergio-teruel ? This is modifying the behavior and I'm not sure this fits in this module.

Reply:
Can you please answer this, @sergio-teruel ?

Reply:
@pedrobaeza I think this is because when using average or FiFo cost price method the standard price can change between moment of order & moment of delivery. When analyzing the margin on deliveries you want to take the cost price at delivery, not at order. The issue I might see is backorders etc.

Reply:
Exactly, you're modifying a price that can be used for several deliveries. IMO, you should leave the one of the purchase confirmation (that's to say, remove this code).

Reply:
Maybe it's better to add purchase_price_delivery and store the standard price there.  Delivery margin calculations should then be based on this purchase price.

Reply:
But you will have the same problem on multiple receptions...

Reply:
When you validate a picking (his stock moves) the product standard price is recomputed (denpending of valuation settings), this is the reason of this line.
And this field is added from sale_margin module, so i think that this module is the correct place...

Reply:
> When you validate a picking (his stock moves) the product standard price is recomputed 

Then `_compute_margin` should be used to avoid currency and uom issues, don't you think?: https://github.com/odoo/odoo/blob/11.0/addons/sale_margin/models/sale_order.py#L14

Anyway, as stated by others, we'd still have the problem with multiple deliveries, backorders, etc. Maybe we could use  `product.price.history` and the mentioned `purchase_price_delivery` so we obtained an accurate margin?